sklearn/mixture/base.py

```diff
@@ -0,0 +1,440 @@
+
+        Parameters
+        ----------
+        X : array-like, shape  (n_samples, n_features)
+
+        resp : array-like, shape (n_samples, n_components)
+        """
+        pass
+
+    def fit(self, X, y=None):
+        """Estimate model parameters with the EM algorithm.
+
+        The method fit the model `n_init` times and set the parameters with
+        which the model has the largest likelihood or lower bound. Within each
+        trial, the method iterates between E-step and M-step for `max_iter`
+        times until the change of likelihood or lower bound is less than
+        `tol`, otherwise, a `ConvergenceWarning` is raised.
+
+        Parameters
+        ----------
+        X : array-like, shape (n_samples, n_features)
+            List of n_features-dimensional data points. Each row
+            corresponds to a single data point.
+
+        Returns
+        -------
+        self
+        """
+        X = _check_X(X, self.n_components)
+        self._check_initial_parameters(X)
+
+        # if we enable warm_start, we will have a unique init
+        do_init = not(self.warm_start and hasattr(self, 'converged_'))
+        n_init = self.n_init if do_init else 1
+
+        max_log_likelihood = -np.infty
+        self.converged_ = False
+
+        for init in range(n_init):
+            # print verbose msg
+            self._print_verbose_msg_init_beg(init)
+
+            if do_init:
+                self._initialize_parameters(X)
+
+            current_log_likelihood = -np.infty
+
+            for n_iter in range(self.max_iter):
+                prev_log_likelihood = current_log_likelihood
+
+                current_log_likelihood, resp = self._e_step(X)
+                change = current_log_likelihood - prev_log_likelihood
+                self._print_verbose_msg_iter_end(n_iter, change)
+
+                if abs(change) < self.tol:
+                    self.converged_ = True
+                    break
+
+                self._m_step(X, resp)
+
+            # print verbose msg
+            self._print_verbose_msg_init_end(current_log_likelihood)
+
+            if current_log_likelihood > max_log_likelihood:
+                # max_log_likelihood is always updated,
+                # since we compute the log-likelihood of the initialization
+                max_log_likelihood = current_log_likelihood
+                best_params = self._get_parameters()
+                self.best_n_iter_ = n_iter
+
+        if not self.converged_:
+            warnings.warn('Initialization %d did not converged. '
+                          'Try different init parameters, '
+                          'or increase n_init, tol '
+                          'or check for degenerate data.'
+                          % (init + 1), ConvergenceWarning)
+            # compute the log-likelihood of the last m-step
+            current_log_likelihood, _ = self._e_step(X)
+
+        self._set_parameters(best_params)
```

Comment:
Also the docstring of the class need to be updated to document the `n_iter_` attribute.